core/src/main/java/kafka/server/AssignmentsManager.java

```diff
@@ -0,0 +1,386 @@
+         */
+        @Override
+        public void handleException(Throwable e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Handles new generated assignments, to be propagated to the controller.
+     * Assignment events may be handled out of order, so for any two assignment
+     * events for the same topic partition, the one with the oldest timestamp is
+     * disregarded.
+     */
+    private class AssignmentEvent extends Event {
+        final long timestampNs;
+        final TopicIdPartition partition;
+        final Uuid dirId;
+        AssignmentEvent(long timestampNs, TopicIdPartition partition, Uuid dirId) {
+            this.timestampNs = timestampNs;
+            this.partition = partition;
+            this.dirId = dirId;
+        }
+        @Override
+        public void run() throws Exception {
+            AssignmentEvent existing = pending.getOrDefault(partition, null);
+            if (existing != null && existing.timestampNs > timestampNs) {
+                if (log.isTraceEnabled()) {
+                    log.trace("Dropping assignment {} because it's older than {}", this, existing);
+                }
+                return;
+            }
+            if (log.isTraceEnabled()) {
+                log.trace("Received new assignment {}", this);
+            }
+            pending.put(partition, this);
+            if (inflight == null || inflight.isEmpty()) {
+                scheduleDispatch();
+            }
+        }
+        @Override
+        public String toString() {
+            return "AssignmentEvent{" +
+                    "timestampNs=" + timestampNs +
+                    ", partition=" + partition +
+                    ", dirId=" + dirId +
+                    '}';
+        }
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            AssignmentEvent that = (AssignmentEvent) o;
+            return timestampNs == that.timestampNs &&
+                    Objects.equals(partition, that.partition) &&
+                    Objects.equals(dirId, that.dirId);
+        }
+        @Override
+        public int hashCode() {
+            return Objects.hash(timestampNs, partition, dirId);
+        }
+    }
+
+    /**
+     * Gathers pending assignments and pushes them to the controller in a {@link AssignReplicasToDirsRequest}.
+     */
+    private class DispatchEvent extends Event {
+        static final String TAG = "dispatch";
+        @Override
+        public void run() throws Exception {
+            if (inflight != null) {
+                throw new IllegalStateException("Bug. Should not be dispatching while there are assignments in flight");
+            }
+            if (pending.isEmpty()) {
+                log.trace("No pending assignments, no-op dispatch");
+                return;
+            }
+            inflight = pending;
```

Comment:
What happens if there are too many Assignment events to handle in a request?

Reply:
We should set a limit. I thought of that but then completely forgot.

What do you think a good limit would be?

The size of request depends not just on the number of assignments, but also on the number of common directories and topics among them, due to the tree-like structure of the request.

If we assume there are 10 directories, for a limit of 2500 assignments per directory, the size of the request can vary in the following way:

```
Topics per dir: 1     Partitions per topic per dir: 2500  Resulting request size: 122KiB
Topics per dir: 50    Partitions per topic per dir: 50    Resulting request size: 131KiB
Topics per dir: 2500  Partitions per topic per dir: 1     Resulting request size: 561KiB
```

With this in mind, I think a limit of 2,5000 assignments per request is reasonable. WDYT?

Reply:
After an offline chat with @pprovenzano we agreed to keep the maximum size under 64kb, so I'm changing the limit to 2250 assignments per request.

```
Topics per dir: 1, Partitions per topic per dir: 225 Resulting request size: 11KiB
Topics per dir: 15, Partitions per topic per dir: 15 Resulting request size: 13KiB
Topics per dir: 225, Partitions per topic per dir: 1 Resulting request size: 50KiB
```